### ISSUE TITLE
Correct answer display for multiple inputs

### ISSUE THREAD

**Issue by samfearn** [maths/moodle-qtype_stack#460]
When a question has multiple inputs, and hence multiple correct answers, then if the question review has 'Right Answer' turned on, the correct answers appear all together at the bottom without any kind of labelling.

I think that students might find this display slightly confusing, as they have to then work out which answer is corresponding to which question they were asked. Obviously, they are just in the order corresponding to the various inputs, but I think it might appear cleaner if different inputs could be labelled as different parts and the answers (and maybe the feedback generated by the PRTs) could then be organised by question part.

**Comment by timhunt** [comment]
Perhaps a better option would be to include the right-answer display in the place where answer validation is usually shown for each input? (Which I think can work, even though input validation can be hidden, there is still a placeholder where we can show the right answer.)

**Comment by samfearn** [comment]
Yes I think that would be a lot clearer rather than all just at the end.

**Comment by sangwinc** [comment]
I agree, and this is something which has annoyed me on occasion.  Should this be in PHP or via javascript?

**Comment by timhunt** [comment]
This should be in PHP. In the renderer.

**Comment by sangwinc** [comment]
Yes, this line
https://github.com/maths/moodle-qtype_stack/blob/master/renderer.php#L440
is where we generate this.  But, how do we get this back into the input validation fields?

What happens in the rare case when a teacher has switched this off? (MCQs are a prime case)

**Comment by timhunt** [comment]
I don't think we should merge this into the validation display.

I think we should show it, styled appropriately (which probably means with the standard Moodle 'feedback' background colour - yellow - text like "Right answer: 1 - x") after the white validation display. Whicever of the right answer display are present should be shown where the [[validation]] placeholder is in the question text.

**Comment by sangwinc** [comment]
Tim,
I'm going to need some advice on how Moodle calls the various question methods! In particular, how we call the code which renders a question with flag that says we should embed the answers.
Chris

**Comment by timhunt** [comment]
Sorry, this question got burried in my inbox.

The flag that says whether we should be displaying the automatically generated information about what the right answer is, is $options->rightanswer, which $options is the question_display_options object passed in to renderer methods like formulation_and_controls.

**Comment by sangwinc** [comment]
Thanks for the tip Tim!  That turned out to be rather easy when you know what to do.
We now have the right information in the right place. There are a few design issues to resolve.
1. The compact validation doesn't work well here.  The "The teacher's answer is..." stuff could be embedded in strange places.
2. We need style.  The div with class=feedback does not do enough to get the standard background colour - yellow.
Any thoughts?

**Comment by timhunt** [comment]
Judging by what other question types have to do, there is no standard class name you can apply to get the standard colouring. (This is bad, we should probably fix that in Moodle core one day.) In the mean time, you will have to add your own CSS rule in STACK.

Tim Hunt
Senior Developer | Information Technology
[cid:image002.jpg@01D634D8.F801DBE0]

**Comment by EJMFarrow** [comment]
I've had a chat with Chris about this and we're now thinking that answer tags in the question text may be the answer e.g. [[answer:ans1]]. If the option to display the right answer is set for a quiz then any tags present will be replaced with the correct answer, otherwise the tags will be stripped out. The answers for any inputs without an answer tag will still appear beneath the question in the normal way. If all the inputs have answer tags, we'll probably need to display a message along the lines of 'Correct answers are displayed in the question' in the normal correct answer location.

Example where fist input doesn't have a tag but the others do:
![iss460](https://github.com/maths/moodle-qtype_stack/assets/31312684/01c55ac1-931b-4a17-9cad-3890420b67ee)

**Comment by EJMFarrow** [comment]
Hi, all. This issue has gone stale again. Are answer tags worth going ahead with?

**Comment by samfearn** [comment]
I like this implementation. Personally I've found that I hide the correct answer in most questions now, and rely on the specific feedback to deliver a correct answer when I want one, so I think it would be good to also get feedback from someone using CorrectAnswer.
